File: harness/bridge/a2a_client.py
from __future__ import annotations

import json
import uuid
from dataclasses import dataclass
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.parse import urljoin, urlparse
from urllib.request import Request, urlopen


class A2AClientError(RuntimeError):
    pass


@dataclass(frozen=True)
class SendAssignmentResult:
    task_id: str
    result: dict[str, Any]


def base_url_from_agent_card_url(agent_card_url: str) -> str:
    parsed = urlparse(agent_card_url)
    suffix = "/.well-known/agent-card.json"
    path = parsed.path
    if path.endswith(suffix):
        path = path[: -len(suffix)] or "/"
    return parsed._replace(path=path, query="", fragment="").geturl().rstrip("/")


def extract_task_id(result: Any) -> str | None:
    if isinstance(result, dict):
        if result.get("id") and result.get("kind") == "task":
            return str(result["id"])
        task = result.get("task")
        if isinstance(task, dict) and task.get("id"):
            return str(task["id"])
        nested = result.get("result")
        if isinstance(nested, dict):
            if nested.get("id"):
                return str(nested["id"])
            nested_task = nested.get("task")
            if isinstance(nested_task, dict) and nested_task.get("id"):
                return str(nested_task["id"])
    return None


class A2AClient:
    def __init__(self, timeout_seconds: float = 30.0) -> None:
        self.timeout_seconds = timeout_seconds

    def send_assignment(
        self,
        *,
        transport: dict[str, Any],
        bearer_token: str,
        push_token: str | None,
        assignment_id: str,
        text: str,
    ) -> SendAssignmentResult:
        endpoint = self._resolve_endpoint(transport, bearer_token)
        params = {
            "configuration": {
                "blocking": False,
                "acceptedOutputModes": ["text/markdown", "application/json"],
            },
            "message": {
                "kind": "message",
                "role": "user",
                "messageId": str(uuid.uuid4()),
                "metadata": {"assignment_id": assignment_id},
                "parts": [{"kind": "text", "text": text}],
            },
            "metadata": {"assignment_id": assignment_id},
        }
        if transport.get("push_url"):
            params["configuration"]["pushNotificationConfig"] = {
                "id": assignment_id,
                "url": transport["push_url"],
                "token": push_token,
                "authentication": {"schemes": ["Bearer"], "credentials": push_token},
            }

        result = self._json_rpc(endpoint, bearer_token, "message/send", params)
        task_id = extract_task_id(result)
        if not task_id:
            raise A2AClientError("A2A message/send response did not include a task id")
        return SendAssignmentResult(task_id=task_id, result=result)

    def cancel_task(self, *, transport: dict[str, Any], bearer_token: str, task_id: str) -> Any:
        endpoint = self._resolve_endpoint(transport, bearer_token)
        return self._json_rpc(endpoint, bearer_token, "tasks/cancel", {"id": task_id})

    def _resolve_endpoint(self, transport: dict[str, Any], bearer_token: str) -> str:
        direct = transport.get("endpoint_url") or transport.get("a2a_url")
        if direct:
            return str(direct)
        agent_card_url = transport.get("agent_card_url")
        if not agent_card_url:
            raise A2AClientError("transport.json needs agent_card_url, endpoint_url, or a2a_url")

        card = self._get_json(str(agent_card_url), bearer_token)
        for interface in card.get("additionalInterfaces") or card.get("additional_interfaces") or []:
            if str(interface.get("transport", "")).upper() in {"JSONRPC", "JSON-RPC"} and interface.get("url"):
                return urljoin(base_url_from_agent_card_url(str(agent_card_url)) + "/", str(interface["url"]))
        if card.get("url"):
            return urljoin(base_url_from_agent_card_url(str(agent_card_url)) + "/", str(card["url"]))
        raise A2AClientError(f"agent card did not expose a JSON-RPC endpoint: {agent_card_url}")

    def _get_json(self, url: str, bearer_token: str) -> dict[str, Any]:
        request = Request(url, headers={"Authorization": f"Bearer {bearer_token}", "Accept": "application/json"})
        try:
            with urlopen(request, timeout=self.timeout_seconds) as response:
                return json.loads(response.read().decode("utf-8"))
        except (HTTPError, URLError, TimeoutError, json.JSONDecodeError) as error:
            raise A2AClientError(f"failed to fetch A2A agent card {url}: {error}") from error

    def _json_rpc(self, endpoint: str, bearer_token: str, method: str, params: dict[str, Any]) -> Any:
        body = json.dumps(
            {"jsonrpc": "2.0", "id": str(uuid.uuid4()), "method": method, "params": params},
            separators=(",", ":"),
        ).encode("utf-8")
        request = Request(
            endpoint,
            data=body,
            method="POST",
            headers={
                "Authorization": f"Bearer {bearer_token}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            },
        )
        try:
            with urlopen(request, timeout=self.timeout_seconds) as response:
                payload = json.loads(response.read().decode("utf-8"))
        except HTTPError as error:
            detail = error.read().decode("utf-8", errors="replace")
            raise A2AClientError(f"A2A {method} failed with HTTP {error.code}: {detail}") from error
        except (URLError, TimeoutError, json.JSONDecodeError) as error:
            raise A2AClientError(f"A2A {method} failed: {error}") from error
        if isinstance(payload, dict) and payload.get("error"):
            raise A2AClientError(f"A2A {method} returned JSON-RPC error: {payload['error']}")
        return payload.get("result") if isinstance(payload, dict) and "result" in payload else payload

File: bridge/src/a2a_client.js
import { randomUUID } from 'node:crypto';
import { ClientFactory, JsonRpcTransport, JsonRpcTransportFactory } from '@a2a-js/sdk/client';

function authFetch(bearerToken, fetchImpl = fetch) {
  return (url, init = {}) => {
    const headers = new Headers(init.headers ?? {});
    headers.set('Authorization', `Bearer ${bearerToken}`);
    return fetchImpl(url, { ...init, headers });
  };
}

function baseUrlFromAgentCardUrl(agentCardUrl) {
  const url = new URL(agentCardUrl);
  if (url.pathname.endsWith('/.well-known/agent-card.json')) {
    url.pathname = url.pathname.slice(0, -'/.well-known/agent-card.json'.length) || '/';
    url.search = '';
    url.hash = '';
  }
  return url.toString().replace(/\/$/, '');
}

function extractTaskId(result) {
  if (result?.id && result?.kind === 'task') return result.id;
  if (result?.task?.id) return result.task.id;
  if (result?.result?.id) return result.result.id;
  if (result?.result?.task?.id) return result.result.task.id;
  return undefined;
}

export class A2AClient {
  constructor({ fetchImpl } = {}) {
    this.fetchImpl = fetchImpl ?? fetch;
  }

  async createClient({ transport, bearerToken }) {
    const authedFetch = authFetch(bearerToken, this.fetchImpl);
    if (transport.agent_card_url) {
      const factory = new ClientFactory({
        transports: [new JsonRpcTransportFactory({ fetchImpl: authedFetch })],
      });
      return factory.createFromUrl(baseUrlFromAgentCardUrl(transport.agent_card_url), '');
    }
    const endpoint = transport.endpoint_url ?? transport.a2a_url;
    if (!endpoint) throw new Error('transport.json needs agent_card_url, endpoint_url, or a2a_url');
    return new JsonRpcTransport({ endpoint, fetchImpl: authedFetch });
  }

  async sendAssignment({ transport, bearerToken, pushToken, assignmentId, text }) {
    const client = await this.createClient({ transport, bearerToken });
    const result = await client.sendMessage({
      configuration: {
        blocking: false,
        acceptedOutputModes: ['text/markdown', 'application/json'],
        pushNotificationConfig: transport.push_url
          ? {
              id: assignmentId,
              url: transport.push_url,
              token: pushToken,
              authentication: { schemes: ['Bearer'], credentials: pushToken },
            }
          : undefined,
      },
      message: {
        kind: 'message',
        role: 'user',
        messageId: randomUUID(),
        metadata: { assignment_id: assignmentId },
        parts: [{ kind: 'text', text }],
      },
      metadata: { assignment_id: assignmentId },
    });
    const taskId = extractTaskId(result);
    if (!taskId) throw new Error(`A2A message/send response did not include a task id`);
    return { taskId, result };
  }

  async cancelTask({ transport, bearerToken, taskId }) {
    const client = await this.createClient({ transport, bearerToken });
    return client.cancelTask({ id: taskId });
  }
}
